synapse/handlers/room_list.py

```diff
@@ -135,114 +147,151 @@ async def _get_public_room_list(
                 This can be (None, None) to indicate the main list, or a particular
                 appservice and network id to use an appservice specific one.
                 Setting to None returns all public rooms across all lists.
-            from_federation: Whether this request originated from a
-                federating server or a client. Used for room filtering.
+            from_client_mxid: A user's MXID if this request came from a registered user.
+            from_remote_server_name: A remote homeserver's server name, if this
+                request came from the federation API.
         """
 
         # Pagination tokens work by storing the room ID sent in the last batch,
         # plus the direction (forwards or backwards). Next batch tokens always
         # go forwards, prev batch tokens always go backwards.
 
+        forwards = True
+        last_joined_members = None
+        last_room_id = None
+        last_module_index = None
         if since_token:
             batch_token = RoomListNextBatch.from_token(since_token)
-
-            bounds: Optional[Tuple[int, str]] = (
-                batch_token.last_joined_members,
-                batch_token.last_room_id,
-            )
             forwards = batch_token.direction_is_forward
-            has_batch_token = True
-        else:
-            bounds = None
-
-            forwards = True
-            has_batch_token = False
+            last_joined_members = batch_token.last_joined_members
+            last_room_id = batch_token.last_room_id
+            last_module_index = batch_token.last_module_index
 
-        # we request one more than wanted to see if there are more pages to come
+        # We request one more than wanted to see if there are more pages to come
         probing_limit = limit + 1 if limit is not None else None
 
-        results = await self.store.get_largest_public_rooms(
+        # We bucket results per joined members number since we want to keep order
+        # per joined members number
+        num_joined_members_buckets: Dict[int, List[PublicRoom]] = {}
+        room_ids_to_module_index: Dict[str, int] = {}
+
+        local_public_rooms = await self.store.get_largest_public_rooms(
             network_tuple,
             search_filter,
             probing_limit,
-            bounds=bounds,
+            bounds=(
+                last_joined_members,
```

Comment:
This may need fixing up for forward pagination.

```
Full public room list:
    [5 members, module None, ...]
  --------------------------------- forward pagination from here: no modules need to be queried for <= 4 members
    [5 members, module None, ...]
    [5 members, module    0, ...]
    [5 members, module    0, ...]
    [5 members, module    1, ...]
  --------------------------------- forward pagination from here: modules None and 0 need to be queried for <= 4 members
    [5 members, module    1, ...]
    [5 members, module    2, ...]
    [5 members, module    2, ...]
    [4 members, module None, ...]
    [4 members, module None, ...]
    [4 members, module    0, ...]
    [4 members, module    0, ...]
    [4 members, module    1, ...]
    [4 members, module    1, ...]
    [4 members, module    2, ...]
    [4 members, module    2, ...]
```